### PR TITLE
Add `jax.scipy.linalg.toeplitz`.

### PR DESCRIPTION
**Pull request by yotarok**
A counterpart for "scipy.linalg.toeplitz" is added in "jax.scipy.linalg".

We recently have a case that requires a Jax counterpart of `tf.linalg.LinearOperatorToeplitz`, and this functionality is in fact a part of SciPy, implemented as `scipy.linalg.toeplitz`.
This also relates to the request in https://github.com/google/jax/issues/10144.

### CODE REVIEW

File: jax/_src/scipy/linalg.py [hunk 1]
```diff
@@ -1032,3 +1032,20 @@ def hessenberg(a: ArrayLike, *, calc_q: bool = False, overwrite_a: bool = False,
     return h, q
   else:
     return h
+
+@_wraps(scipy.linalg.toeplitz)
+def toeplitz(c, r=None):
```
Comment (jakevdp) [inline jax/_src/scipy/linalg.py:1037]: Add type annotations:
```python
def toeplitz(c: ArrayLike, r: Optional[ArrayLike] = None) -> Array:
  ...
```
Also, I think the type checker will require you to use different variable names for the flattened version of these below.

File: tests/linalg_test.py [hunk 2]
```diff
@@ -1679,5 +1679,42 @@ def testSchurBatching(self, shape, dtype):
       Ts, Ss = vmap(lax.linalg.schur)(args)
       self.assertAllClose(reconstruct(Ss, Ts), args, atol=1e-4)
 
+  @jtu.sample_product(
+      cshape=[(4,), (8,), (3, 7)],
+      cdtype=float_types + complex_types,
+      rshape=[(3,), (7,), (2, 1, 4)],
```
Comment (jakevdp) [inline tests/linalg_test.py:1685]: You might also test shapes `()` and `(0,)` to make sure you're covering all corner cases.

**Review by jakevdp — CHANGES_REQUESTED** [review]
Looks great! Just a few comments below and we can get this in.

Also, please add the new function to the list here so it will show up in the HTML docs: https://github.com/google/jax/blob/main/docs/jax.scipy.rst#jaxscipylinalg

File: tests/linalg_test.py [hunk 3]
```diff
@@ -1679,5 +1679,42 @@ def testSchurBatching(self, shape, dtype):
...
+      rdtype=float_types + complex_types)
+  def testToeplitzConstrcution(self, rshape, rdtype, cshape, cdtype):
+      rng = jtu.rand_default(self.rng())
+      args_maker = lambda: [rng(cshape, cdtype), rng(rshape, rdtype)]
+      promotion = jax.config.jax_numpy_dtype_promotion
+      if {rdtype, cdtype} == {np.float32, np.complex64}:
+        # Test that default complex-promotion behavior is matching with osp.
+        promotion = 'standard'
+      with jax.numpy_dtype_promotion(promotion):
```
[the remark this answers is not in the document]
Comment (yotarok) [reply tests/linalg_test.py]: done

File: jax/_src/scipy/linalg.py [hunk 4]
```diff
@@ -1032,3 +1032,20 @@ def hessenberg(a: ArrayLike, *, calc_q: bool = False, overwrite_a: bool = False,
...
+  if r is None:
```
[the remark this answers is not in the document]
Comment (yotarok) [reply jax/_src/scipy/linalg.py:1039]: done
thanks, I didn't know this is an API standard: https://github.com/google/jax/issues/7737

File: jax/_src/scipy/linalg.py [hunk 5]
```diff
@@ -1032,3 +1032,20 @@ def hessenberg(a: ArrayLike, *, calc_q: bool = False, overwrite_a: bool = False,
...
+    r = jnp.conjugate(jnp.asarray(c))
+
+  c = jnp.asarray(c).flatten()
+  r = jnp.asarray(r).flatten()
+
+  ncols, = c.shape
+  nrows, = r.shape
+  nelems = ncols + nrows - 1
```
[the remark this answers is not in the document]
Comment (yotarok) [reply jax/_src/scipy/linalg.py:1054]: done. added a short-cut for such corner cases.

File: jax/_src/scipy/linalg.py (hunk 1 above)
Comment (yotarok) [reply jax/_src/scipy/linalg.py]: done

File: tests/linalg_test.py (hunk 2 above)
Comment (yotarok) [reply tests/linalg_test.py]: done

**Review by jakevdp — APPROVED** [review]
Looks great, thanks!

**Comment by yotarok** [comment]
Looks like the numerical error is severe when TPUs/ GPUs are used because `lax.conv_general_dilated_patches` is used.
So, I switched to use `vmap` and `np.roll` that are slightly slower but should be accurate.

**Comment by jakevdp** [comment]
I suspect `vmap` of `roll` is going to be much less efficient than the original approach, due to how it's implemented. Would you like to change back to the convolution-based approach along with adjusting the matmul precision?

**Comment by yotarok** [comment]
The profiling results with TPU were:
- lax.conv_general_dilated_patches: ~9ms
- roll + vmap: ~15ms
- lax.conv_general_dilated_patches + Precision.HIGHEST: ~12ms
(with configuaration of `r == c`, `r.shape = (1024, 256)` and `toeplitz` is called with `vmap`).

Confirmed that with `Precision.HIGHEST`, `assertAllClose` in the test passes with the default tolerance parameters with TPU/ GPU.

**Comment by yotarok** [comment]
I noticed that integers are not tested.  So, I added integer tests and some skip conditions because CuDNN doesn't support integer convolution except for int8.
PTAL again.

File: tests/linalg_test.py [hunk 6]
```diff
@@ -1679,5 +1680,62 @@ def testSchurBatching(self, shape, dtype):
       Ts, Ss = vmap(lax.linalg.schur)(args)
       self.assertAllClose(reconstruct(Ss, Ts), args, atol=1e-4)
 
+  @jtu.sample_product(
```
Comment (jakevdp) [inline tests/linalg_test.py:1683]: These tests should probably go under `class ScipyLinalgTestCase` rather than `class LaxLinalgTestCase`

File: tests/linalg_test.py [hunk 7]
```diff
@@ -1679,5 +1680,62 @@ def testSchurBatching(self, shape, dtype):
...
+      cshape=[(), (4,), (8,), (3, 7), (0, 5, 1)],
+      cdtype=float_types + complex_types,
+      rshape=[(), (3,), (7,), (2, 1, 4), (19, 0)],
+      rdtype=float_types + complex_types + int_types)
+  def testToeplitzConstrcution(self, rshape, rdtype, cshape, cdtype):
+      if ((rdtype in [np.float64, np.complex128]
```
Comment (jakevdp) [inline tests/linalg_test.py:1689]: Here and throughout, please use two-space indentations.,

**Review by jakevdp — CHANGES_REQUESTED** [review]
Last couple minor comments - thanks!

File: tests/linalg_test.py (hunk 6 above)
Comment (yotarok) [reply tests/linalg_test.py]: moved.

File: tests/linalg_test.py (hunk 7 above)
Comment (yotarok) [reply tests/linalg_test.py]: done.
What's the style recommendation for hanging indents?
For `_src/scipy/linalg.py`, I followed Google-style, i.e. 2 spaces for normal indents, and 4 spaces for hanging indents.
However, `tests/linalg_test.py` uses 2 spaces also for hanging indents, so I followed that file-local convention only for the test.

**Comment by yotarok** [comment]
Thank you for the comments, I updated the PR accordingly.

**Review by jakevdp — APPROVED** [review]
Thanks!
